src/main/java/net/neoforged/neoforge/capabilities/RegisterCapabilitiesEvent.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.ArrayList;
+import java.util.Objects;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.ItemLike;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.event.IModBusEvent;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Fired to register capability providers at an appropriate time.
+ */
+public class RegisterCapabilitiesEvent extends Event implements IModBusEvent {
+    // TODO: cleanup once eventbus changes are in
+    @ApiStatus.Internal // Public due to eventbus reasons.
+    public RegisterCapabilitiesEvent() {}
+
+    // BLOCKS
+
+    /**
+     * Register a capability provider for some blocks.
+     * <p>
+     * <b>If a previously returned capability is not valid anymore, or if a new capability is available,
+     * {@link Level#invalidateCapabilities(BlockPos)} MUST be called to notify the caches.</b>
+     * See {@link IBlockCapabilityProvider} for details.
+     */
+    public <T, C> void registerBlock(BlockCapability<T, C> capability, IBlockCapabilityProvider<T, C> provider, Block... blocks) {
+        Objects.requireNonNull(provider);
+        // Probably a programmer error
+        if (blocks.length == 0)
+            throw new IllegalArgumentException("Must register at least one block");
+
+        for (Block block : blocks) {
+            Objects.requireNonNull(block);
+            capability.providers.computeIfAbsent(block, b -> new ArrayList<>()).add(provider);
+        }
+    }
+
+    /**
+     * Register a capability provider for a block entity type.
+     * <p>
+     * <b>If a previously returned capability is not valid anymore, or if a new capability is available,
+     * {@link Level#invalidateCapabilities(BlockPos)} MUST be called to notify the caches.</b>
+     * See {@link IBlockCapabilityProvider} for details.
+     */
+    public <T, C, BE extends BlockEntity> void registerBlockEntity(BlockCapability<T, C> capability, BlockEntityType<BE> blockEntityType, ICapabilityProvider<? super BE, C, T> provider) {
+        Objects.requireNonNull(provider);
+
+        IBlockCapabilityProvider<T, C> adaptedProvider = (level, pos, state, blockEntity, context) -> {
+            // TODO: not sure that the != type check is really needed; but just in case?
```

Comment:
It is. A block can technically have different BE types at different points in time